### DISCUSSION TITLE
Add support for hypergraphs and hyperedges

### DISCUSSION THREAD

**Discussion by scottcode** [networkx/networkx#5117] [Ideas]
Would the networkx development community consider adding support for [hypergraphs](https://en.wikipedia.org/wiki/Hypergraph)? A hypergraph is a graph that allows a type of edge that connects more than just two nodes (hyperedges). This seems like an exciting area that is gaining traction in media and literature (see [How Big Data Carried Graph Theory Into New Dimensions](https://www.quantamagazine.org/how-big-data-carried-graph-theory-into-new-dimensions-20210819/)). I am working on datasets that would benefit from such representation.

Simple multipartite (heterogeneous) graphs can be represented in Networkx, but as long as they have only binary edges then they fail to capture important truths about the underlying data.

To illustrate this, consider a dataset about sports teams, the stadiums where they have played, and against whom they have played. The dataset might have triplets of `(home_team, away_team, stadium)`. You could create a tripartite graph. While you might be able to reconstruct who has played with whom for regular stadiums that belong to just one team, and where that team is always the home team. However, this becomes problematic for championship games, which tend to be hosted in special stadiums and are associated with multiple "home" teams. Simple [projection](https://en.wikipedia.org/wiki/Bipartite_network_projection) onto team node types could lead you to infer that many pairs had played each other at that stadium, even though in reality they hadn't.

Hyperedges disambiguate such situations and would pave the way for networkx users to explore the new avenues of graph happening in the [article](https://www.quantamagazine.org/how-big-data-carried-graph-theory-into-new-dimensions-20210819/) I shared above.

**Comment by dschult** [comment]
We have looked at Hypergraphs as an interesting extension of networkx over the years. But we haven't found a compelling data structure or even a general set of algorithms around which to base a software package. Indeed, a HyperGraph is a set of nodes and a collection of subsets of those nodes. That's very general. 

Suggestions are welcome. :}

**Comment by MridulS** [comment]
It would definitely be nice to have Hypergraphs as part of networkx!

I remember fiddling around with hypergraphs couple of years back and to reconcile the networkx data structure and Hypergraphs I ended "projecting" hyperedges as complete graphs and repurposing `MultiGraphs` using the `keys` as an "edge layer". Something like
``` python
In [1]: import networkx as nx

In [2]: from collections import defaultdict

In [3]: from itertools import combinations

In [4]: class HyperGraph(nx.MultiGraph):
   ...:     def __init__(self, *args, **kwargs):
   ...:         self._hyper_adj = defaultdict(dict)
   ...:         self._layer = 0
   ...:         super().__init__(*args, **kwargs)
   ...:     def add_edge(self, nodes, layer=None):
   ...:         if layer is None:
   ...:             layer = self._layer
   ...:             self._layer += 1
   ...:         for node in nodes:
   ...:             if node not in self:
   ...:                 self.add_node(node)
   ...:         for u, v in combinations(nodes, 2):
   ...:             super().add_edge(u, v)
   ...:             if layer in self._hyper_adj[u]:
   ...:                 self._hyper_adj[u][layer].add(v)
   ...:             else:
   ...:                 self._hyper_adj[u][layer] = set([v])
   ...:             if layer in self._hyper_adj[v]:
   ...:                 self._hyper_adj[v][layer].add(u)
   ...:             else:
   ...:                 self._hyper_adj[v][layer] = set([u])
   ...:     def add_edges_from(self, iter_of_nodes):
   ...:         for nodes in iter_of_nodes:
   ...:             self.add_edge(nodes)
   ...:

In [5]: G = HyperGraph(); G.add_edges_from([(1, 2, 3), (2, 3), (3, 5, 6), (4,)])

In [6]: G._hyper_adj # the "hyper" adjacency dictionary which tells us which nodes are part of which hyperedge.
Out[6]:
defaultdict(dict,
            {1: {0: {2, 3}},
             2: {0: {1, 3}, 1: {3}},
             3: {0: {1, 2}, 1: {2}, 2: {5, 6}},
             5: {2: {3, 6}},
             6: {2: {3, 5}}})
``` 
But this seems pretty inefficient to me and it also loses the flexibility of using MultiGraph inside HyperGraphs. Maybe we just need to go another level deeper in the dict-o-dict-o-dict structure 🙃.

There seems to be some other packages like https://github.com/pnnl/HyperNetX, we could probably take some inspiration.

**Comment by scottcode** [comment]
@MridulS I've seen HyperNetX but haven't tried it. Might be able to learn from their abstractions and design choices. 

@dschult : One method that would be useful associated with hypergraph representation would be projection, analogous to bipartite projection. Given a multi-partite graph with hyperedges and a target node type, form connections between nodes of that type only when they appeared in hyperedges that shared the same nodes of the other (dropped) node types.
